### DISCUSSION TITLE
Daten aus Tasmota / ESPeasy JSON String meiner Wärmepumpe

### DISCUSSION THREAD

**Discussion by LurchiStromberg** [camueller/SmartApplianceEnabler#408] [Installation und Betrieb]
Hallo!

Ich nutze zum Auslesen des Stromverbrauchs meiner Wärmepumpe ESPeasy das Modul Stromcount. Ist am Ende nichts anderes, als Tasmota. Ich habe auch sOnOff Devices, die mit SAE super funktionieren.

Leider hängt es bei mir nun bei der Auswertung des JSON Strings, weil ich leider nur sehr begrenzte Fähigkeiten in Sachen programmieren habe.

Der String sieht wie folgt aus und wird durch Aufruf der Seite http://"ip des Device"/json] abgerufen.

`"Sensors":[
{
"TaskValues": [
{"ValueNumber":1,
"Name":"Count",
"NrDecimals":2,
"Value":0.00
},
{"ValueNumber":2,
"Name":"Total",
"NrDecimals":2,
"Value":4187.00
},
{"ValueNumber":3,
"Name":"Time",
"NrDecimals":2,
"Value":77901.00
}],
"DataAcquisition": [
{"Controller":1,
"IDX":0,
"Enabled":"true"
},
{"Controller":2,
"IDX":0,
"Enabled":"false"
},
{"Controller":3,
"IDX":0,
"Enabled":"false"
}],
"TaskInterval":10,
"Type":"Generic - Pulse counter",
"TaskName":"Watt",
"TaskDeviceNumber":3,
"TaskEnabled":"true",
"TaskNumber":1
}
],
"TTL":10000
}`

Der ausschlaggebende Wert ist Time:

```
{"ValueNumber":3,
"Name":"Time",
"NrDecimals":2,
"Value":77901.00

```
Die Formel für den aktuellen Verbrauch in Watt ist: 3600 / Time * 1000 (3600 / 77901 * 1000=46,21 Watt)

Ich komme mit dem Pfad nicht klar (3 x gleiche Namen, nur die ValueNumber ist anders) und der Umrechnungsformel in SAE. URL ist ja noch simpel, aber was kommt bei Daten, Pfad und regex rein?

Kann mir dabei bitte jemand helfen, wie ich diese Werte aus dem JSON in SAE einbinde?
Danke!

**Comment by camueller** [comment]
Du kannst nur Datenquellen verwenden, die entweder den Zählerstand oder die Leistung als Zahlenwert bereitstellen. Diese Wert nach irgendeine Formel zu berechnen, wie Du oben angeben hast, ist nicht möglich.

**Reply by LurchiStromberg** [reply]
> Du kannst nur Datenquellen verwenden, die entweder den Zählerstand oder die Leistung als Zahlenwert bereitstellen. Diese Wert nach irgendeine Formel zu berechnen, wie Du oben angeben hast, ist nicht möglich.

Danke, das du mir bei dem "Problem" hilfst.

OK, die Umrechnung könnte ich im ESPeasy Modul machen habe ich grade gesehen.

Bleibt trotzdem noch das Problem des Auslesens...
Wie bekomme ich aus dem JSON:

{"ValueNumber":3,
"Name":"Time",
"NrDecimals":2,
"Value":77901.00

den Wert Value":77901.00, also die 77901.00 (würde ich dann wie gesagt schon in ESPeasy umrechnen, also stände da 46.21) in SAE?

**Comment by camueller** [comment]
Habe gerade die Doku nochmal angepasst: https://github.com/camueller/SmartApplianceEnabler/blob/master/doc/ValueExtraction_DE.md

**Reply by LurchiStromberg** [reply]
> Habe gerade die Doku nochmal angepasst: https://github.com/camueller/SmartApplianceEnabler/blob/master/doc/ValueExtraction_DE.md

Super, damit lässt sich wirklich was anfangen :)
Allerdings möchte der Wechselrichter meinen ESPeasy immer noch nicht.

Hier mal der komplette JSON String:
`{"System":{
"Build":20106,
"Git Build":"",
"System Libraries":"ESP82xx Core 9c56ed1f, NONOS SDK 2.2.2-dev(38a443e), LWIP: 2.1.2 PUYA support",
"Plugin Count":46,
"Plugin Description":"[Normal]",
"Local Time":"2022-12-23 19:53:22",
"Unit Number":0,
"Unit Name":"Stromcount",
"Uptime":2259,
"Last Boot Cause":"Manual reboot",
"Reset Reason":"Software/System restart",
"Load":5.16,
"Load LC":4174,
"CPU Eco Mode":"false",
"Heap Max Free Block":16888,
"Heap Fragmentation":5,
"Free RAM":17712
},
"WiFi":{
"Hostname":"Stromcount",
"IP Config":"Static",
"IP Address":"192.168.2.70",
"IP Subnet":"255.255.255.0",
"Gateway":"192.168.2.1",
"STA MAC":"80:7D:3A:78:63:75",
"DNS 1":"8.8.8.8",
"DNS 2":"(IP unset)",
"SSID":"7490Fritzbox",
"BSSID":"28:D1:27:4E:1B:F6",
"Channel":4,
"Connected msec":135534155,
"Last Disconnect Reason":1,
"Last Disconnect Reason str":"(1) Unspecified",
"Number Reconnects":0,
"Force WiFi B/G":"false",
"Restart WiFi Lost Conn":"false",
"Force WiFi No Sleep":"false",
"Periodical send Gratuitous ARP":"false",
"Connection Failure Threshold":0,
"RSSI":-64
},
"Sensors":[
{
"TaskValues": [
{"ValueNumber":1,
"Name":"Count",
"NrDecimals":2,
"Value":0.00
},
{"ValueNumber":2,
"Name":"Total",
"NrDecimals":2,
"Value":6088.00
},
{"ValueNumber":3,
"Name":"Time",
"NrDecimals":2,
"Value":29.51
}],
"DataAcquisition": [
{"Controller":1,
"IDX":0,
"Enabled":"true"
},
{"Controller":2,
"IDX":0,
"Enabled":"false"
},
{"Controller":3,
"IDX":0,
"Enabled":"false"
}],
"TaskInterval":10,
"Type":"Generic - Pulse counter",
"TaskName":"Watt",
"TaskDeviceNumber":3,
"TaskEnabled":"true",
"TaskNumber":1
}
],
"TTL":10000
}`

Daraus habe ich folgenden Pfad extrahiert:
$.Sensors[0].TaskValues[2].Value und bei Pfad eingetragen.
Die Umrechnung macht nun ESPeasy, Wert ist also aktuell 29,51
Allerdings bekomme ich vom Wechselrichter beim hinzufügen des Geräts einen Fehler...
Kannst du dir da einen Reim draus machen?

[https://ibb.co/yymFrWH](url)

**Reply by camueller** [reply]
Beim "Hinzufügen" meinst Du sicher das Sunny Portal. Der Wechselrichter spielt hier keine Rolle.
Dein Link funktioneirt leider nicht.

Jedenfalls hat das Extrahieren von Werten aus dem JSON nichts mit dem Hinzufügen von Geräten zum Sunny Portal zu tun. Wenn letzteres nicht geht und Du die [Hinweise](https://github.com/camueller/SmartApplianceEnabler/blob/master/doc/SunnyPortal_DE.md) beachtet hast, musst Du Dich wohl an den SMA Support wenden.

**Reply by LurchiStromberg** [reply]
Oh sorry....
https://ibb.co/yymFrWH
Einfach den Link kopieren, hier schein Github was durcheinander zu würfeln.

Also hat das nichts mit dem Wert im Eigentlichen zu tun sondern mit dem Gerät, welches hinzugefügt wird?
Aber in meinem Fall ist es doch nichts anderes als bei Tasmota...
Eine Zahl aus einem JSON String....

Hänge da im Moment auf dem Schlauch.
Hatte letztes Wochenende erst noch einen sOnOff POW hinzugefügt, ohne Probleme

**Reply by camueller** [reply]
Unten steht: "Sie können noch 11 Geräte hinzufügen", d.h. Du hast vermutlich bereits 12 steuerbare Geräte hinzugefügt. Das ist das Maixmum an steuerbaren Geräten. Wenn einige Geräte nicht steuerbar sind, kann man bis zu 22 Geräten im Sunny Portal haben.

**Reply by LurchiStromberg** [reply]
Steuern tue ich keines der Geräte, sind alles Tasmota sOnOff POWs.

Verstehe auch grade nicht, warum die Fehlermeldung kommt...

**Reply by LurchiStromberg** [reply]
> Unten steht: "Sie können noch 11 Geräte hinzufügen", d.h. Du hast vermutlich bereits 12 steuerbare Geräte hinzugefügt.

Verdammt, dass hat etwas gedauert bis ich das kapiert habe.
JA! Sämtliche SAE haben den Verbrauchertyp "Direkte Kommunikation"

Da ich keine der sOnOff POWs schalte, wäre "nur messen" vollkommen ausreichend (und somit hätte ich ja auch 24 Geräte)

Nur wo kann ich das umstellen? Oder werden SAE Geräte mit Verbrauchertyp "Direkte Kommunikation" hinzugefügt?

EDIT:
OK, das wird bei SAE in "Typ" festgelegt?
Da kann man natürlich nicht einstellen "nur messen", oder?
